content/docs/guides/crossguard/core-concepts.md

```diff
@@ -0,0 +1,96 @@
+---
+title: Core Concepts
+linktitle: Core Concepts
+
+menu:
+  userguides:
+    parent: crossguard
+    identifier: crossguard-core-concepts
+---
+{{% crossguard-preview %}}
+
+## Policy
+
+A Policy contains specific logic you would like to enforce. For example, you may want to restrict the creation of public S3 buckets or you may disallow resource provisioning without tags. You can refer to other examples [here](https://github.com/pulumi/examples/tree/master/policy-packs).
+
+Policies are written as validation functions that are evaluated against all resources in your Pulumi stack. If the validation function calls `reportViolation`, the associated resource will be considered in violation of the policy. `reportViolation` can be called multiple times to report multiple violations.
+
+Policies validation functions are executed during `pulumi preview` and `pulumi update`, asserting that cloud resource definitions comply with the policy immediately before they are created or updated.
+
+During `preview`, every resource is checked for policy violations, which are batched up into a final report. During the `update`, the first policy violation will halt the deployment.
```

Comment:
> every resource is checked for policy violations, which are batched up into a final report.

I don't believe this is actually true. I think there is a bug open for this, but I believe we fail the preview immediately rather than batching them up for the end. (I couldn't find a bug or comment in the code, perhaps @justinvp can keep me honest.)

... we _want_ it to be true, but we just aren't there yet.

Reply:
Checked with Justin and we are indeed batching for previews.

Reply:
There was a TODO in the code, but it appears to have been out-of-date, and the comment was deleted by Cyrus as part of some refactoring a week ago.

Reply:
https://github.com/pulumi/pulumi/commit/1908a18d203f8a857248c31c13f3404342c76038#diff-152cf3031402283e2b7aa998a44043ccL282-L283

```diff
-			// TODO(hausdorff): Batch up failures and report them all at once during preview. This code here
- 			// will cause them to fail eagerly, and stop the plan immediately.
```